### PR TITLE
8202617: javadoc generates broken links to undocumented (e.g. private) members

### PR DESCRIPTION
**Pull request by nizarbenalla**
Please review this patch to prevent links to private and package-private members to be generated.
The bug happens when you link to private/package-private members, and javadoc used to generated links to them (assuming they were inherited because the holder is unreachable).

Taking the code path I changed is very rare, as it only used by 4 anchors in 4 classes in all the JDK.
```
if (refSignature.trim().startsWith("#") &&
                    ! (utils.isPublic(containing) || utils.isLinkable(containing))
```

The classes that used it are `StringBuilder`/`StringBuffer` with `#append(java.lang.String)` and `ZipEntry`/`ZipOutputStream` with `#CENHDR`

I've expanded the test to check whether the links are created when they should be.

The generated documentation before and after the change are identical.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8202617: javadoc generates broken links to undocumented (e.g. private) members`

### Issue
 * [JDK-8202617](https://bugs.openjdk.org/browse/JDK-8202617): javadoc generates broken links to undocumented (e.g. private) members (**Bug** - P3)

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21802/head:pull/21802` \
`$ git checkout pull/21802`

Update a local copy of the PR: \
`$ git checkout pull/21802` \
`$ git pull https://git.openjdk.org/jdk.git pull/21802/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21802`

View PR using the GUI difftool: \
`$ git pr show -t 21802`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21802.diff">https://git.openjdk.org/jdk/pull/21802.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21802#issuecomment-2449667859)
</details>

### CODE REVIEW

**Comment by hns** [comment]
In the PR description you quote `StringBuilder#append(java.lang.String)` and `ZipEntry#CENHDR` as examples for this problem, but those are both public members. In the current JDK docs, I can only find a [link for the latter][CENHDR]  which is actually working.

[CENHDR]: https://download.java.net/java/early_access/jdk24/docs/api/java.base/java/util/zip/ZipEntry.html#CENHDR

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/LinkTaglet.java [hunk 1]
```diff
@@ -249,7 +249,8 @@ Content linkSeeReferenceOutput(Element holder,
                 }
             }
             if (refSignature.trim().startsWith("#") &&
-                    ! (utils.isPublic(containing) || utils.isLinkable(containing))) {
+                    ! (utils.isPublic(containing) || utils.isLinkable(containing)) &&
+                    ! (utils.isPrivate(refMem) ||utils.isPackagePrivate(refMem))) {
```
Comment (hns) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/LinkTaglet.java:253]: Shouldn't this take into account whether we are running with -private or -package options? Ideally using utils.isIncluded(Element) would take care of this and give the right answer in all cases, although I haven't tried.

Comment (nizarbenalla) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/taglets/LinkTaglet.java]: I observed no changes when building javadoc using `-private` or `-package` before/after the change, I don't think this code path is taken in those cases.
`utils.isIncluded(refMem)`/`utils.isIncluded(containing)` is always false, I wasn't sure how to use it. (Which element should I check?)

File: test/langtools/jdk/javadoc/doclet/5093723/T5093723.java [hunk 2]
```diff
@@ -23,30 +24,73 @@
...
+                """, """
+                package p;
+                class UndocumentedClass {
+                    void method() {}
+                    public void publicMethod() {}
+                    protected void protectedMethod() {}
+                    private void privateMethod() {}
+                }
+                """);
+
```
Comment (hns) [inline test/langtools/jdk/javadoc/doclet/5093723/T5093723.java:78]: Kudos for "modernizing" the test!

**Review by hns — CHANGES_REQUESTED** [review]
Sorry, my initial feedback was wrong. `utils.isIncluded` is indeed not what we want to use, because `refMem` and `containing` are in fact not included (which is why we want to link to the inherited documentation in the current class). 

So I had to refresh my knowledge of [doclet terminology], because what we want to know is whether the referenced member is _selected_ rather than _included_. In other words, we want to check if selection control allows the member to be documented in the current class.

I'm adding an inline comment with the change I suggest to take this into account.

[doclet terminology]: https://docs.oracle.com/en/java/javase/23/docs/api/jdk.javadoc/jdk/javadoc/doclet/package-summary.html#terminology-heading

File: test/langtools/jdk/javadoc/doclet/5093723/T5093723.java (hunk 2 above)
Comment (nizarbenalla) [reply test/langtools/jdk/javadoc/doclet/5093723/T5093723.java:78]: Thanks!

**Comment by nizarbenalla** [comment]
I updated the code using the suggestion by Hannes and a small check in the test to make sure the warnings are emitted.
Here is the javadoc output when trying to build Javadoc for `DocumentedClass` and `UndocumentedClass` with the latest patch
```
Generating ./out1/allpackages-index.html...
Generating ./out1/index-all.html...
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:26: warning: reference not accessible: UndocumentedClass
  /** {@link #method} */
      ^
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:28: warning: reference not accessible: UndocumentedClass
  /** {@link #publicMethod} */
      ^
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:30: warning: reference not accessible: UndocumentedClass
  /** {@link #protectedMethod} */
      ^
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:32: warning: reference not accessible: UndocumentedClass
  /** {@link #privateMethod} */
      ^
Generating ./out1/search.html...
Generating ./out1/index.html...

```
Before (in [432ba68](https://github.com/openjdk/jdk/pull/21802/commits/432ba68602b7d5df8abbd918f11e6b9c615823a1)), the output was the following. I didn't notice it was missing a couple of warnings.
```
Building index for all classes...
Generating ./out1/allpackages-index.html...
Generating ./out1/index-all.html...
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:28: warning: reference not accessible: UndocumentedClass
  /** {@link #publicMethod} */
      ^
test/langtools/jdk/javadoc/doclet/5093723/DocumentedClass.java:30: warning: reference not accessible: UndocumentedClass
  /** {@link #protectedMethod} */
      ^
Generating ./out1/search.html...
```

Here is the [Javadoc](https://cr.openjdk.org/~nbenalla/GeneratedDocs/8202617/out1/package-summary.html) after running the `javadoc` tool on `DocumentedClass` and `UndocumentedClass`

**Comment by hns** [comment]
Unfortunately, there is something fishy about the warnings. There shouldn't be a "reference not accessible" warning for the public and protected method links, because these are in fact accessible in the current type. It turns out these warnings are generated when condition `htmlWriter instanceof ClassWriter cw` is `false`, i.e. when we are generating an index or summary page. 

One way of fixing this would be to rewrite the 3-branch if statement in lines 257-268 in the following way:

                if (utils.configuration.docEnv.isSelected(refMem)) {
                    if (htmlWriter instanceof ClassWriter cw) {
                        containing = cw.getTypeElement();
                    }
                } else {
                    reportWarning.accept("doclet.link.see.reference_not_accessible",
                            new Object[] { refMem });
                }

This retains the quirk in the current code that the member is only linked if we are in the class writer. We could get the containing class by getting the type element for `holder`, but IMO the current behaviour is acceptable for such a corner case.

Note that I changed the second argument for the warning to the actual member reference, since that is what is unaccessible. Also note that the original `else` branch is not needed anymore as the warning is always about the member not being accessible/selected.

**Comment by nizarbenalla** [comment]
I no longer see the warnings on public and protected methods after the latest change. 
I thought the warnings were fine because those methods belonged to an Undocumented class (even if they are inherited).
Here is the [Javadoc](https://cr.openjdk.org/~nbenalla/GeneratedDocs/8202617-0f10f99/out1/package-summary.html) after the latest changes.

**Review by hns — APPROVED** [review]
Thanks Nizar, looks good now!

**Comment by nizarbenalla** [comment]
Thanks for the help on this

/integrate

**Pull request closed without merging**